engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java

```diff
@@ -204,32 +208,45 @@ protected void onActorCloseRequested() {
         () ->
             actor.runOnCompletion(
                 streamProcessor.getLastWrittenPositionAsync(),
-                (writtenPosition, ex1) -> {
-                  if (ex1 == null) {
-                    actor.runOnCompletion(
-                        streamProcessor.getLastProcessedPositionAsync(),
-                        (processedPosition, ex2) -> {
-                          if (ex2 == null) {
-                            enforceSnapshotCreation(writtenPosition, processedPosition);
-                            close();
-                            future.complete(null);
-                          } else {
-                            LOG.error(ERROR_MSG_ON_RESOLVE_PROCESSED_POS, ex2);
-                            close();
-                            future.completeExceptionally(ex2);
-                          }
-                        });
-
-                  } else {
-                    LOG.error(ERROR_MSG_ON_RESOLVE_WRITTEN_POS, ex1);
-                    close();
-                    future.completeExceptionally(ex1);
-                  }
-                }));
+                (writtenPosition, ex1) -> handleLastWrittenPosition(future, writtenPosition, ex1)));
 
     return future;
   }
 
+  private void handleLastWrittenPosition(
+      final CompletableActorFuture<Void> future, final Long writtenPosition, final Throwable ex1) {
+    if (ex1 != null) {
+      LOG.error(ERROR_MSG_ON_RESOLVE_WRITTEN_POS, ex1);
+      close();
+      future.completeExceptionally(ex1);
+    } else if (writtenPosition == null) {
+      LOG.warn(WARN_NULL_WRITTEN_POS);
+    } else {
+      actor.runOnCompletion(
+          streamProcessor.getLastProcessedPositionAsync(),
+          (processedPosition, ex2) ->
+              handleLastProcessedPosition(future, writtenPosition, processedPosition, ex2));
+    }
+  }
+
+  private void handleLastProcessedPosition(
+      final CompletableActorFuture<Void> future,
+      final Long writtenPosition,
+      final Long processedPosition,
+      final Throwable ex2) {
+    if (ex2 != null) {
+      LOG.error(ERROR_MSG_ON_RESOLVE_PROCESSED_POS, ex2);
+      close();
+      future.completeExceptionally(ex2);
+    } else if (processedPosition == null) {
+      LOG.warn(WARN_NULL_PROCESSED_POS);
```

Comment:
Here as well, this is a problem on closing right?

Reply:
My bad, I forgot to complete it